File: fontend/lib/online-class.ts
/**
 * Online Class API client — mirrors backend/app/schemas/online_class.py.
 *
 * One surface serves both consoles: teacher endpoints are guarded by the
 * teaching scope, student endpoints by the caller's enrollment. The live
 * classroom itself rides a WebSocket (see hooks/use-live-room.ts).
 */

import { API_BASE_URL, getAccessToken } from "./auth";
import { leadershipCall, queryString } from "./principal";

const call = <T>(path: string, init: RequestInit = {}): Promise<T> =>
  leadershipCall<T>("online-classes", path, init, "OnlineClassAPIError");

const jsonInit = (method: string, body: unknown): RequestInit => ({
  method,
  headers: { "Content-Type": "application/json" },
  body: JSON.stringify(body),
});

/** WS URL for the live room; the short-lived JWT rides in the query string. */
export function liveRoomUrl(classId: string): string {
  const base = API_BASE_URL.replace(/^http/, "ws");
  return `${base}/api/v1/online-classes/${classId}/live?token=${encodeURIComponent(getAccessToken() ?? "")}`;
}

export function uploadUrl(classId: string, kind: "files" | "recording"): string {
  return `${API_BASE_URL}/api/v1/online-classes/${classId}/${kind}`;
}

export function fileHref(url: string): string {
  return url.startsWith("http") ? url : `${API_BASE_URL}${url}`;
}

// ── Shapes ───────────────────────────────────────────────────────────────────

export interface TeachingAssignmentOption {
  subject_id: string;
  subject_code: string;
  subject_name: string;
  class_id: string;
  class_name: string;
}

export interface SetupSlot {
  id: string;
  class_id: string;
  class_name: string;
  subject_id: string | null;
  subject_name: string | null;
  period_number: number;
  start_time: string;
  end_time: string;
}

export interface OnlineClassSetupOptions {
  assignments: TeachingAssignmentOption[];
  today_slots: SetupSlot[];
}

export interface OnlineClassCreate {
  class_id: string;
  subject_id: string;
  topic: string;
  scheduled_at?: string | null;
  duration_minutes: number;
  allow_join: boolean;
  recording_enabled: boolean;
  timetable_slot_id?: string | null;
}

export interface OnlineClassRow {
  id: string;
  class_id: string;
  class_name: string;
  subject_id: string;
  subject_code: string;
  subject_name: string;
  teacher_id: string;
  teacher_name: string;
  topic: string;
  mode: "SCHEDULED" | "INSTANT";
  status: "SCHEDULED" | "LIVE" | "COMPLETED" | "CANCELLED";
  scheduled_at: string | null;
  duration_minutes: number;
  allow_join: boolean;
  recording_enabled: boolean;
  recording_url: string | null;
  started_at: string | null;
  ended_at: string | null;
  created_at: string;
  participant_count: number;
}

export interface OnlineParticipantRow {
  student_id: string;
  student_name: string;
  roll_number: string | null;
  waiting_since: string;
  joined_at: string | null;
  left_at: string | null;
  duration_seconds: number;
  attendance_status: string | null;
  hand_raised_at: string | null;
  is_online: boolean;
}

export interface OnlineFileRow {
  id: string;
  uploader_name: string;
  file_name: string;
  url: string;
  file_size_bytes: number;
  mime_type: string;
  created_at: string;
}

export interface OnlineMessageRow {
  id: string;
  sender_id: string;
  sender_name: string;
  sender_role: string;
  body: string;
  created_at: string;
}

export interface OnlineClassDetail extends OnlineClassRow {
  roster_size: number;
  participants: OnlineParticipantRow[];
  files: OnlineFileRow[];
  /** Populated on the student view only. */
  join_state?: JoinState;
}

export interface OnlineClassPage {
  total: number;
  limit: number;
  offset: number;
  items: OnlineClassRow[];
}

export interface OnlineAttendanceRow {
  student_id: string;
  student_name: string;
  roll_number: string | null;
  joined_at: string | null;
  left_at: string | null;
  duration_seconds: number;
  percent: number | null;
  attendance_status: "PRESENT" | "LATE" | "ABSENT";
}

export interface OnlineAttendanceReport {
  class_id: string;
  class_name: string;
  subject_name: string;
  topic: string;
  started_at: string | null;
  ended_at: string | null;
  duration_seconds: number;
  present_min_percent: number;
  late_min_percent: number;
  totals_present: number;
  totals_late: number;
  totals_absent: number;
  rows: OnlineAttendanceRow[];
}

export type JoinState = "NOT_ELIGIBLE" | "UPCOMING" | "JOINABLE" | "WAITING" | "IN_CLASS" | "ENDED";

export interface StudentOnlineClassRow extends OnlineClassRow {
  join_state: JoinState;
}

export interface StudentOnlineClassList {
  today: StudentOnlineClassRow[];
  upcoming: StudentOnlineClassRow[];
  past: StudentOnlineClassRow[];
}

// ── Teacher calls ────────────────────────────────────────────────────────────

export const fetchSetupOptions = () => call<OnlineClassSetupOptions>("/setup-options");
export const scheduleOnlineClass = (payload: OnlineClassCreate) =>
  call<OnlineClassRow>("", jsonInit("POST", payload));
export const startInstantClass = (payload: OnlineClassCreate) =>
  call<OnlineClassRow>("/instant", jsonInit("POST", payload));
export const fetchTeacherOnlineClasses = (status?: string) => call<OnlineClassPage>(queryString({ status }));
export const fetchOnlineClassDetail = (id: string) => call<OnlineClassDetail>(`/${id}`);
export const startOnlineClass = (id: string) => call<OnlineClassRow>(`/${id}/start`, { method: "POST" });
export const endOnlineClass = (id: string) => call<OnlineAttendanceReport>(`/${id}/end`, { method: "POST" });
export const cancelOnlineClass = (id: string) => call<OnlineClassRow>(`/${id}/cancel`, { method: "POST" });
export const patchOnlineClass = (id: string, payload: { allow_join?: boolean; recording_enabled?: boolean }) =>
  call<OnlineClassRow>(`/${id}`, jsonInit("PATCH", payload));
export const admitStudent = (id: string, studentId: string) =>
  call<OnlineClassDetail>(`/${id}/participants/${studentId}/admit`, { method: "POST" });
export const admitAllStudents = (id: string) => call<OnlineClassDetail>(`/${id}/admit-all`, { method: "POST" });
export const removeStudentFromClass = (id: string, studentId: string) =>
  call<OnlineClassDetail>(`/${id}/participants/${studentId}/remove`, { method: "POST" });
export const fetchAttendanceReport = (id: string) => call<OnlineAttendanceReport>(`/${id}/attendance`);
export const fetchTeacherChatHistory = (id: string) => call<OnlineMessageRow[]>(`/${id}/messages`);

// ── Student calls ────────────────────────────────────────────────────────────

export const fetchMyOnlineClasses = () => call<StudentOnlineClassList>("/my/classes");
export const fetchStudentClassView = (id: string) => call<OnlineClassDetail>(`/${id}/student-view`);
export const joinOnlineClass = (id: string) => call<StudentOnlineClassRow>(`/${id}/join`, { method: "POST" });
export const leaveOnlineClass = (id: string) => call<StudentOnlineClassRow>(`/${id}/leave`, { method: "POST" });
export const fetchStudentChatHistory = (id: string) => call<OnlineMessageRow[]>(`/${id}/student/messages`);
export const fetchClassMaterials = (id: string) => call<OnlineFileRow[]>(`/${id}/student/files`);

// ── Uploads (multipart, token in header) ─────────────────────────────────────

export async function uploadClassFile(classId: string, kind: "files" | "recording", file: File): Promise<void> {
  const form = new FormData();
  form.append("file", file);
  const res = await fetch(uploadUrl(classId, kind), {
    method: "POST",
    headers: { Authorization: `Bearer ${getAccessToken()}` },
    body: form,
  });
  if (!res.ok) throw new Error(`Upload failed (${res.status})`);
}

File: backend/app/schemas/online_class.py
"""Wire contracts for the Online Class module.

Snake_case payloads, matching the other institution console clients
(``fontend/lib/online-class.ts`` mirrors these one-to-one).
"""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.common import APIResponse
from app.schemas.teacher import TeacherScheduleSlot, TeachingAssignment


# ── Setup options (teacher) ───────────────────────────────────────────────────


class OnlineClassSetupOptions(BaseModel):
    assignments: list[TeachingAssignment]
    today_slots: list[TeacherScheduleSlot]


# ── Create / update payloads ──────────────────────────────────────────────────


class OnlineClassCreate(BaseModel):
    class_id: uuid.UUID
    subject_id: uuid.UUID
    topic: str = Field(min_length=1, max_length=255)
    scheduled_at: datetime | None = None
    duration_minutes: int = Field(default=60, ge=5, le=480)
    allow_join: bool = True
    recording_enabled: bool = False
    timetable_slot_id: uuid.UUID | None = None


class OnlineClassUpdate(BaseModel):
    allow_join: bool | None = None
    recording_enabled: bool | None = None


# ── Rows ──────────────────────────────────────────────────────────────────────


class OnlineClassRow(BaseModel):
    id: uuid.UUID
    class_id: uuid.UUID
    class_name: str
    subject_id: uuid.UUID
    subject_code: str
    subject_name: str
    teacher_id: uuid.UUID
    teacher_name: str
    topic: str
    mode: str
    status: str
    scheduled_at: datetime | None = None
    duration_minutes: int
    allow_join: bool
    recording_enabled: bool
    recording_url: str | None = None
    started_at: datetime | None = None
    ended_at: datetime | None = None
    created_at: datetime
    participant_count: int = 0


class OnlineParticipantRow(BaseModel):
    student_id: uuid.UUID
    student_name: str
    roll_number: str | None = None
    waiting_since: datetime
    joined_at: datetime | None = None
    left_at: datetime | None = None
    duration_seconds: int = 0
    attendance_status: str | None = None
    hand_raised_at: datetime | None = None
    is_online: bool = False


class OnlineFileRow(BaseModel):
    id: uuid.UUID
    uploader_name: str
    file_name: str
    url: str
    file_size_bytes: int
    mime_type: str
    created_at: datetime


class OnlineMessageRow(BaseModel):
    id: uuid.UUID
    sender_id: uuid.UUID
    sender_name: str
    sender_role: str
    body: str
    created_at: datetime


class OnlineMessageIn(BaseModel):
    body: str = Field(min_length=1, max_length=1000)


class OnlineClassDetail(OnlineClassRow):
    roster_size: int = 0
    participants: list[OnlineParticipantRow] = []
    files: list[OnlineFileRow] = []
    join_state: str | None = None  # populated for the student view


class OnlineClassPage(BaseModel):
    total: int
    limit: int
    offset: int
    items: list[OnlineClassRow]


# ── Attendance (the automatic report) ────────────────────────────────────────


class OnlineAttendanceRow(BaseModel):
    student_id: uuid.UUID
    student_name: str
    roll_number: str | None = None
    joined_at: datetime | None = None
    left_at: datetime | None = None
    duration_seconds: int = 0
    percent: float | None = None
    attendance_status: str


class OnlineAttendanceReport(BaseModel):
    class_id: str
    class_name: str
    subject_name: str
    topic: str
    started_at: datetime | None = None
    ended_at: datetime | None = None
    duration_seconds: int
    present_min_percent: float
    late_min_percent: float
    totals_present: int
    totals_late: int
    totals_absent: int
    rows: list[OnlineAttendanceRow]


# ── Student console ───────────────────────────────────────────────────────────


class StudentOnlineClassRow(OnlineClassRow):
    join_state: str  # NOT_ELIGIBLE | UPCOMING | JOINABLE | WAITING | IN_CLASS | ENDED


class StudentOnlineClassList(BaseModel):
    today: list[StudentOnlineClassRow]
    upcoming: list[StudentOnlineClassRow]
    past: list[StudentOnlineClassRow]


# ── Envelope aliases ──────────────────────────────────────────────────────────

APIResponseOnlineClassSetupOptions = APIResponse[OnlineClassSetupOptions]
APIResponseOnlineClass = APIResponse[OnlineClassRow]
APIResponseOnlineClassDetail = APIResponse[OnlineClassDetail]
APIResponseOnlineClassPage = APIResponse[OnlineClassPage]
APIResponseOnlineAttendanceReport = APIResponse[OnlineAttendanceReport]
APIResponseOnlineMessages = APIResponse[list[OnlineMessageRow]]
APIResponseOnlineMessage = APIResponse[OnlineMessageRow]
APIResponseOnlineFiles = APIResponse[list[OnlineFileRow]]
APIResponseOnlineFile = APIResponse[OnlineFileRow]
APIResponseStudentOnlineClasses = APIResponse[StudentOnlineClassList]
